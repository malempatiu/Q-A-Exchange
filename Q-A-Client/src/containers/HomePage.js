import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchQuestions, removeError} from '../actions/actions';
import { Link } from 'react-router-dom';

class HomePage extends Component {
    state = {
        text: ''
    }
    componentDidMount() {
        this.props.removeError();
        this.props.fetchQuestions();
    };
    handleUserSearch = (e) => {
        this.setState({ text: e.target.value });
    };
    render() {
        const { questions, error } = this.props;
        let newQuestions;
        if (this.state.text.length > 0) {
            newQuestions = questions.filter((question) => {
                return question.text.toLowerCase().includes(this.state.text.toLowerCase());
            });
        } else {
            newQuestions = questions;
        }
        return (
            <div className="conatiner homepage">
                <div className="jumbotron text-center">
                    <h1>Welcome to Q-A-Exchange</h1>
                    <p className="lead">View all top questions and answers</p>
                    <Link className="btn btn-info text-white" to="/addquestion">Ask Question</Link>
                </div>
                <div className="container text-center w-75">
                    <input type="text" className="form-control searchInput" placeholder="search questions....." onChange={this.handleUserSearch} style={{ margin: '8px' }} />
                </div>
                <div className="container">
                    {error && <div className="alert alert-danger">{error}</div>}
                    {newQuestions.map((question) => {
                        return <div key={question._id} className="row">
                            <div className="col-lg-12 col-md-12 col-sm-12">
                                <div className="card">
                                    <Link to={`/question/${question._id}`} id={question._id} className="card-body text-primary">
                                        {question.text}
                                    </Link>
                                </div>
                            </div>
                        </div>
                    })}
                </div>
            </div>
        );
    }
};

const mapStateToProps = (state) => {
    return {
        questions: state.Questions.questions,
        error: state.Errors.error
    };
}

export default connect(mapStateToProps, { fetchQuestions, removeError })(HomePage);