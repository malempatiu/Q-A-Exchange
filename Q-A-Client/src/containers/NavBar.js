import React, { Component } from 'react';
import { Link } from 'react-router-dom';

class NavBar extends Component {
    render() {
        return (
            <nav className="navbar navbar-expand-lg navbar-dark bg-dark sticky-top">
                <div className="container">
                    <h1 className="navbar-brand text-white font-weight-bold">Q-A-Exchange</h1>
                    <div className="navbar-toggler" data-toggle="collapse" data-target="#navbarMenu">
                        <span className="navbar-toggler-icon"></span>
                    </div>
                    <div className="collapse navbar-collapse" id="navbarMenu">
                        <div className="navbar-nav ml-auto">
                            <Link className="nav-item nav-link btn btn-info text-white" to="/">Home</Link>
                            <Link className="nav-item nav-link btn btn-info text-white" to="/addquestion">Ask Question</Link>
                        </div>
                    </div>
                </div>
            </nav>
        );
    };
};

export default NavBar;