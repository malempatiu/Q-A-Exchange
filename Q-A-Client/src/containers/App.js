import React, { Component } from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { combineReducers, createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Questions from '../reducers/questions-reducer';
import Errors from '../reducers/errors-reducer';
import QuestionAndAnswer from '../reducers/question-answer-reducer';
import NavBar from './NavBar';
import HomePage from './HomePage';
import CreateQuestion from '../components/CreateQuestion';
import QuestionAnswer from '../components/QuestionAnswer';

const store = createStore(
  combineReducers({
    Questions,
    QuestionAndAnswer,
    Errors
  }),
  compose(applyMiddleware(thunk))
);

class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <div className="app-cont">
          <BrowserRouter>
            <div>
              <NavBar />
              <Switch>
                <Route exact path="/" component={HomePage} />
                <Route exact path="/addquestion" component={CreateQuestion} />
                <Route exact path="/question/:id" component={QuestionAnswer} />
              </Switch>
            </div>
          </BrowserRouter>
        </div>
      </Provider>
    );
  }
}

export default App;

