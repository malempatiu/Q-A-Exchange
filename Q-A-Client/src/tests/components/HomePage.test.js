import React from 'react';
import { HomePage } from '../../containers/HomePage';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const questions = [
    {
        _id: "5b8d55fcfac57133747fc5c4",
        text: "What is React?"
    },
    {
        _id: "5b8d55fcfac57133747fc5c5",
        text: "What is the difference between state and props in React?",
    }
]
describe('HOMEPAGE CONTAINER', () => {
    let removeErrorSpy;
    let fetchQuestionsSpy;
    let wrapper;
    beforeEach(() => {
        removeErrorSpy = jest.fn();
        fetchQuestionsSpy = jest.fn();
        wrapper = shallow(<HomePage questions={questions} removeError={removeErrorSpy} fetchQuestions={fetchQuestionsSpy} />);
    });
    it('should render homepage', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('input')).toHaveLength(1);
        expect(wrapper.find('Link')).toHaveLength(3);
    });
    it('should set state text', () => {
        const value = 'React';
        wrapper.find('input').simulate('change', {
            target: { value }
        });
        expect(wrapper.state('text')).toBe(value);
    });
});