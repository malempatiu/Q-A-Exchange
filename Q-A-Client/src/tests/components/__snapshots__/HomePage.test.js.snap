// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOMEPAGE CONTAINER should render homepage 1`] = `
<div
  className="conatiner homepage"
>
  <div
    className="jumbotron text-center"
  >
    <h1>
      Welcome to Q-A-Exchange
    </h1>
    <p
      className="lead"
    >
      View all top questions and answers
    </p>
    <Link
      className="btn btn-info text-white"
      replace={false}
      to="/addquestion"
    >
      Ask Question
    </Link>
  </div>
  <div
    className="container text-center w-75"
  >
    <input
      className="form-control searchInput"
      onChange={[Function]}
      placeholder="search questions....."
      style={
        Object {
          "margin": "8px",
        }
      }
      type="text"
    />
  </div>
  <div
    className="container"
  >
    <div
      className="row"
      key="5b8d55fcfac57133747fc5c4"
    >
      <div
        className="col-lg-12 col-md-12 col-sm-12"
      >
        <div
          className="card"
        >
          <Link
            className="card-body text-primary"
            id="5b8d55fcfac57133747fc5c4"
            replace={false}
            to="/question/5b8d55fcfac57133747fc5c4"
          >
            What is React?
          </Link>
        </div>
      </div>
    </div>
    <div
      className="row"
      key="5b8d55fcfac57133747fc5c5"
    >
      <div
        className="col-lg-12 col-md-12 col-sm-12"
      >
        <div
          className="card"
        >
          <Link
            className="card-body text-primary"
            id="5b8d55fcfac57133747fc5c5"
            replace={false}
            to="/question/5b8d55fcfac57133747fc5c5"
          >
            What is the difference between state and props in React?
          </Link>
        </div>
      </div>
    </div>
  </div>
</div>
`;
